launchpad/src/launchpad.rs

```diff
@@ -1,9 +1,500 @@
+        require!(
+            !self.blacklist().contains(&caller),
+            "You have been put into the blacklist and may not confirm tickets"
+        );
+
+        let ticket_payment_token = self.ticket_payment_token().get();
+        let ticket_price = self.ticket_price().get();
+        let total_ticket_price = Self::BigUint::from(nr_tickets_to_confirm) * ticket_price;
+
+        require!(
+            payment_token == ticket_payment_token,
+            "Wrong payment token used"
+        );
+        require!(payment_amount == total_ticket_price, "Wrong amount sent");
+
+        let (first_ticket_id, last_ticket_id) = self.try_get_ticket_range(&caller)?;
+        let nr_tickets = last_ticket_id - first_ticket_id + 1;
+
+        require!(
+            nr_tickets >= nr_tickets_to_confirm,
+            "Trying to confirm too many tickets"
+        );
+
+        let current_generation = self.current_generation().get();
+        let mut actual_confirmed_tickets = 0;
+
+        for ticket_id in first_ticket_id..=last_ticket_id {
+            let ticket_status = self.ticket_status().get(ticket_id);
+            if !ticket_status.is_winning(current_generation) {
+                continue;
+            }
+
+            self.set_ticket_status(ticket_id, TicketStatus::Confirmed);
+            actual_confirmed_tickets += 1;
+
+            if actual_confirmed_tickets == nr_tickets_to_confirm {
+                break;
+            }
+        }
+
+        require!(
+            actual_confirmed_tickets == nr_tickets_to_confirm,
+            "Couldn't confirm all tickets"
+        );
+
+        self.total_confirmed_tickets()
+            .update(|confirmed| *confirmed += actual_confirmed_tickets);
+
+        Ok(())
+    }
+
+    #[endpoint(selectNewWinners)]
+    fn select_new_winners(&self) -> SCResult<OperationCompletionStatus> {
+        self.require_stage(LaunchStage::SelectNewWinners)?;
+
+        let (prev_first_winning_ticket_position, prev_last_winning_ticket_position) =
+            self.winning_tickets_range().get();
+        let winning_tickets =
+            prev_last_winning_ticket_position - prev_first_winning_ticket_position + 1;
+        let confirmed_tickets = self.total_confirmed_tickets().get();
+        let remaining_tickets = winning_tickets - confirmed_tickets;
+
+        let new_first_winning_ticket_position = prev_first_winning_ticket_position + 1;
+        let new_last_winning_ticket_position =
+            new_first_winning_ticket_position + remaining_tickets - 1;
+        let last_valid_ticket_id = self.get_total_tickets();
+
+        require!(
+            new_last_winning_ticket_position <= last_valid_ticket_id,
+            "Cannot select new winners, reached end of range"
+        );
+
+        let next_generation = self.current_generation().get() + 1;
+        let mut current_ticket_position = 0;
+
+        let run_result = self.run_while_it_has_gas(|gas_op| match gas_op {
+            GasOp::Load(op) => match op {
```

Comment:
I would take Load operation as a separate function before run_while_it_has_gas.
And GasOp:Completed outside - after run_while_it_has_gas.